rbd: add `group snap info` command

ajarr opened the pull request:
See https://github.com/ceph/ceph/pull/57797#issuecomment-2233172495 for example outputs

<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [X] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

ajarr commented:
TODOs:
- Tests
- Add C API and python bindings for group_snap_info
- tackle any backward compatibility issue with adding `id` in `rbd_group_snap_info_t` structure

ajarr commented:
Request for feedback on the current `rbd group snap info` command output. See https://github.com/ceph/ceph/pull/57797#issue-2326608527

I think we should also show some info about the `.group` snapshots in the `group snap info` output. This way the user can quickly fetch the details of a `.group` snapshot to perform an operation, e.g. cloning a .group snapshot.

Option 1:
```
$ rbd group snap info data/gp@snap1
rbd group snapshot 'snap1':
	id: 104edb3201c6
	state: ok
	image snapshots:
		 data/img1@group.3_10407fbbaf15_104edb3201c6 (snapid: 4)
		 data/img2@group.3_10407fbbaf15_104edb3201c6 (snapid: 3)
                 data/img3@group.3_10407fbbaf15_104edb3201c6 (snapid: 5)
```

Option 2 that includes a table:
```
$ rbd group snap info data/gp@snap1
rbd group snapshot 'snap1':
	id: 104edb3201c6
	state: ok
	image snapshots:
	        IMAGESPEC   SNAPID               SNAPNAME
		 data/img1       4           group.3_10407fbbaf15_104edb3201c6
		 data/img2       3           group.3_10407fbbaf15_104edb3201c6
                 data/img3       5           group.3_10407fbbaf15_104edb3201c6
```

idryomov commented:
> I think we should also show some info about the `.group` snapshots in the `group snap info` output. This way the user can quickly fetch the details of a `.group` snapshot to perform an operation, e.g. cloning a .group snapshot.

Agreed.  I like option 1 better and would propose the following enhancement, recognizing that `.group` snapshots for a particular group snapshot are going to have the same name across all images:
```
$ rbd group snap info data/gp@snap1
rbd group snapshot 'snap1':
	id: 104edb3201c6
	state: ok
        image snap: .group.3_10407fbbaf15_104edb3201c6
	images:
		 data/img1 (snapid: 4)
		 data/img2 (snapid: 3)
                 data/img3 (snapid: 5)
```

sunilangadi2 commented:
> > I think we should also show some info about the `.group` snapshots in the `group snap info` output. This way the user can quickly fetch the details of a `.group` snapshot to perform an operation, e.g. cloning a .group snapshot.
> 
> Agreed. I like option 1 better and would propose the following enhancement, recognizing that `.group` snapshots for a particular group snapshot are going to have the same name across all images:
> 
> ```
> $ rbd group snap info data/gp@snap1
> rbd group snapshot 'snap1':
> 	id: 104edb3201c6
> 	state: ok
>         image snap: .group.3_10407fbbaf15_104edb3201c6
> 	images:
> 		 data/img1 (snapid: 4)
> 		 data/img2 (snapid: 3)
>                  data/img3 (snapid: 5)
> ```

Just thought in terms of metrics perspective as well in future if we need these information in the form of metrics
keeping image snapshots with `IMAGESPEC`    `SNAPID` header would be useful.

idryomov commented:
> Just thought in terms of metrics perspective as well in future if we need these information in the form of metrics
> keeping image snapshots with `IMAGESPEC` `SNAPID` header would be useful.

@sunilangadi2 By "in the form of metrics", are you to parsing the output in scripts?  If so, formatted output (e.g. `--format json`) should be used for that.

idryomov commented on src/tracing/librbd.tp:
```diff
@@ -2873,6 +2873,28 @@ TRACEPOINT_EVENT(librbd, group_snap_list_exit,
     )
 )
 
+TRACEPOINT_EVENT(librbd, group_snap_info_enter,
```
Same as in the other PR -- I wouldn't bother with this tracepoint stuff.

idryomov commented on src/tools/rbd/action/Group.cc:
```diff
@@ -902,6 +1014,10 @@ Shell::Action action_group_snap_list(
   "", &get_group_snap_list_arguments, &execute_group_snap_list);
+Shell::Action action_group_snap_info(
+  {"group", "snap", "info"}, {},
+  "Show information of a group snapshot.",
```
```suggestion
  "Show information about a group snapshot.",
```

idryomov commented on src/librbd/librbd.cc:
```diff
@@ -265,6 +265,7 @@ void group_info_cpp_to_c(const librbd::group_info_t &cpp_info,
 
 void group_snap_info_cpp_to_c(const librbd::group_snap_info_t &cpp_info,
 			      rbd_group_snap_info_t *c_info) {
+  c_info->id = strdup(cpp_info.id.c_str());
```
Need to update `rbd_group_snap_list_cleanup()` accordingly.

idryomov commented on src/librbd/api/Group.cc:
```diff
@@ -1223,13 +1223,60 @@ int Group<I>::snap_list(librados::IoCtx& group_ioctx, const char *group_name,
+  }
+
+  cls::rbd::GroupSnapshot *cls_group_snap = nullptr;
+  for (auto &i : cls_group_snaps) {
```
```suggestion
  for (const auto &i : cls_group_snaps) {
```

idryomov commented on src/librbd/api/Group.cc:
```diff
@@ -1223,13 +1223,60 @@ int Group<I>::snap_list(librados::IoCtx& group_ioctx, const char *group_name,
+  group_snap->name = cls_group_snap->name;
+  group_snap->state = static_cast<group_snap_state_t>(cls_group_snap->state);
+
+  for (auto &i: cls_group_snap->snaps) {
```
```suggestion
  for (const auto &i : cls_group_snap->snaps) {
```

idryomov commented on src/librbd/api/Group.cc:
```diff
@@ -1223,13 +1223,60 @@ int Group<I>::snap_list(librados::IoCtx& group_ioctx, const char *group_name,
+
+  cls::rbd::GroupSnapshot *cls_group_snap = nullptr;
+  for (auto &i : cls_group_snaps) {
+    if (i.name == string(snap_name)) {
```
No need to create a temporary `std::string`:
```suggestion
    if (i.name == snap_name)) {
```

ajarr commented on src/include/rbd/librbd.h:
```diff
@@ -250,6 +250,13 @@ typedef enum {
+} rbd_group_image_snap_info_t;
+
+typedef struct {
+  char *id;
```
Decided to create new APIs and data structure to preserve backwards compatibility

ajarr commented:
`group snap info` outputs:

CLI
```
 $ rbd group snap info data/gp@snap1
rbd group snapshot 'snap1':
	id: 104ccb3288f3
	state: ok
	image snap: .group.3_103e77dd2be9_104ccb3288f3
	images:
		data/img1 (snap id: 3)
		data/img2 (snap id: 4)
		data/img3 (snap id: 5)
```

CLI -- formated JSON
```
$ ./bin/rbd group snap info data/gp@snap1 --format=json
{"id":"104ccb3288f3","name":"snap1","state":"ok","images":[{"pool_name":"data","namespace":"","image_name":"img1","snap_name":".group.3_103e77dd2be9_104ccb3288f3","snap_id":3},{"pool_name":"data","namespace":"","image_name":"img2","snap_name":".group.3_103e77dd2be9_104ccb3288f3","snap_id":4},{"pool_name":"data","namespace":"","image_name":"img3","snap_name":".group.3_103e77dd2be9_104ccb3288f3","snap_id":5}]}
```

Python
```
>>> import rados; import rbd; rc = rados.Rados(conffile='/home/rraja/git/ceph-3/build/ceph.conf'); rc.connect(); ioctx = rc.open_ioctx('data'); gc = rbd.Group(ioctx, 'gp')
>>> print(gc.show_snap('snap1'))
{'id': '104ccb3288f3', 'name': 'snap1', 'state': 1, 'images': [{'pool_id': 3, 'namespace': '', 'image_name': 'img1', 'snap_name': '.group.3_103e77dd2be9_104ccb3288f3', 'snap_id': 3}, {'pool_id': 3, 'namespace': '', 'image_name': 'img2', 'snap_name': '.group.3_103e77dd2be9_104ccb3288f3', 'snap_id': 4}, {'pool_id': 3, 'namespace': '', 'image_name': 'img3', 'snap_name': '.group.3_103e77dd2be9_104ccb3288f3', 'snap_id': 5}]}
>>> ioctx.close(); rc.shutdown()
```

nbalacha commented on src/librbd/librbd.cc:
```diff
@@ -1422,6 +1429,13 @@ namespace librbd {
+  int RBD::group_snap_list2(IoCtx& group_ioctx, const char *group_name,
+                            std::vector<group_snap_info_v2_t> *snaps)
+  {
+
```
Nit: Unnecessary blank line.

nbalacha commented on src/librbd/api/Group.cc:
```diff
@@ -1223,9 +1223,29 @@ int Group<I>::snap_list(librados::IoCtx& group_ioctx, const char *group_name,
+}
+
+template <typename I>
+int Group<I>::snap_list(librados::IoCtx& group_ioctx, const char *group_name,
```
Would it be better to rename this to snap_list2 ?

idryomov replied:
I think the strategy that is used for other APIs to have a single internal handler which is called for both old and new API and have the massaging needed for the old API performed in `src/librbd/librbd.cc`.  So I'd change the existing `Group::snap_list()` to return `group_snap_info2_t` and get rid of the overload that is introduced here.

idryomov commented:
> Python

```
{
    'id': '104ccb3288f3',
    'name': 'snap1',
    'state': 1,
    'images': [
        {
            'pool_id': 3,
            'namespace': '',
            'image_name': 'img1',
            'snap_name': '.group.3_103e77dd2be9_104ccb3288f3',
            'snap_id': 3
        },
        {
            'pool_id': 3,
            'namespace': '',
            'image_name': 'img2',
            'snap_name': '.group.3_103e77dd2be9_104ccb3288f3',
            'snap_id': 4
        },
        {
            'pool_id': 3,
            'namespace': '',
            'image_name': 'img3',
            'snap_name': '.group.3_103e77dd2be9_104ccb3288f3',
            'snap_id': 5
        }
    ]
}
```
Same comments as on the formated CLI output.

idryomov commented:
Can you also paste `rbd group snap ls` output -- it includes the snap ID now, right?

ajarr commented:
> Can you also paste `rbd group snap ls` output -- it includes the snap ID now, right?

Yes

```
$ ./bin/rbd group snap ls data/gp
ID                       NAME   STATUS
10549dbdd25   snap1          ok
105617e77e71  snap2         ok
10581057509a  snap3         ok


$ ./bin/rbd group snap ls data/gp --format json --pretty-format
[
    {
        "id": "10549dbdd25",
        "snapshot": "snap1",
        "state": "ok"
    },
    {
        "id": "105617e77e71",
        "snapshot": "snap2",
        "state": "ok"
    },
    {
        "id": "10581057509a",
        "snapshot": "snap3",
        "state": "ok"
    }
]
```

idryomov commented on src/include/rbd/librbd.h:
```diff
@@ -1476,6 +1490,20 @@ CEPH_RBD_API int rbd_group_snap_list(rados_ioctx_t group_p,
+CEPH_RBD_API  int  rbd_group_snap_info(
+    rados_ioctx_t group_p, const char *group_name, const char *snap_name,
+    rbd_group_snap_info_v2_t *group_snap,
+    rbd_group_image_snap_info_t *image_snaps, size_t *num_image_snaps);
```
I'm not sure I'm following the rationale for having `rbd_group_snap_list2()` return just the id and the name of the snapshot and leaving the list of images that participated in the snapshot to this API.  All of this information is stored in the same `GroupSnapshot` structure in RADOS, so it gets fetched and decoded anyway.

Have you considered defining `rbd_group_snap_info2_t` as something like
```
typedef struct {
  char *id;
  rbd_snap_namespace_type_t unused;
  char *name;
  rbd_group_snap_state_t state;
  char *image_snap_name;
  size_t image_snaps_count;
  rbd_group_image_snap_info_t *image_snaps;
} rbd_group_snap_info2_t;
```
and returning that from both `rbd_group_snap_list2()` for all snapshots and `rbd_group_snap_get_info()` for a particular snapshot?

ajarr replied:
Nice suggestion. I've implemented this.

idryomov commented on src/include/rbd/librbd.h:
```diff
@@ -1476,6 +1490,20 @@ CEPH_RBD_API int rbd_group_snap_list(rados_ioctx_t group_p,
+                                      size_t *num_entries);
+CEPH_RBD_API int rbd_group_snap_list2_cleanup(rbd_group_snap_info_v2_t *snaps,
+                                              size_t num_entries);
+CEPH_RBD_API  int  rbd_group_snap_info(
```
Let's include a verb in the name:
```suggestion
CEPH_RBD_API int rbd_group_snap_get_info(
```

sunilangadi2 commented:
@ajarr Is it expected to have `"snap_name": ".group.3_103e77dd2be9_104ccb3288f3"` starting with `.`?

ajarr commented:
**Fetching information about a group snapshot using CLI**

```
$ rbd group snap info data/gp@snap1
rbd group snapshot 'snap1':
	id: 104e6d3699e
	state: ok
	image snap: .group.3_103a6510642_104e6d3699e
	images:
		data/img1 (snap id: 3)
		data/img2 (snap id: 4)
		data/img3 (snap id: 5)

$ rbd group snap info data/gp@snap1 --format=json --pretty-format
{
    "id": "104e6d3699e",
    "name": "snap1",
    "image_snap_name": ".group.3_103a6510642_104e6d3699e",
    "state": "ok",
    "images": [
        {
            "pool_name": "data",
            "image_name": "img1",
            "snap_id": 3
        },
        {
            "pool_name": "data",
            "image_name": "img2",
            "snap_id": 4
        },
        {
            "pool_name": "data",
            "image_name": "img3",
            "snap_id": 5
        }
    ]
}
```

**Fetching information about a group snapshot using python API**
```
>>> import rados; import rbd; rc = rados.Rados(conffile='/home/rraja/git/ceph-3/build/ceph.conf'); rc.connect(); ioctx = rc.open_ioctx('data'); gc = rbd.Group(ioctx, 'gp')
>>> print(json.dumps(gc.get_snap_info('snap1'), indent=2))
{
  "id": "104e6d3699e",
  "name": "snap1",
  "state": 1,
  "image_snap_name": ".group.3_103a6510642_104e6d3699e",
  "images": [
    {
      "pool_id": 3,
      "snap_id": 3,
      "image_name": "img1"
    },
    {
      "pool_id": 3,
      "snap_id": 4,
      "image_name": "img2"
    },
    {
      "pool_id": 3,
      "snap_id": 5,
      "image_name": "img3"
    }
  ]
}
```

**Listing group snapshots using CLI**
```
$ rbd group snap ls data/gp
ID            NAME   STATUS
104e6d3699e   snap1      ok
1050aa0ea3dc  snap2      ok
1052641a00bb  snap3      ok
$ rbd group snap ls data/gp --format=json --pretty-format
[
    {
        "id": "104e6d3699e",
        "snapshot": "snap1",
        "state": "ok"
    },
    {
        "id": "1050aa0ea3dc",
        "snapshot": "snap2",
        "state": "ok"
    },
    {
        "id": "1052641a00bb",
        "snapshot": "snap3",
        "state": "ok"
    }
]
```

**Listing group snapshots using Python interface**
```
>>> for i in gc.list_snaps():
...     print(json.dumps((i), indent=2))
... 
{
  "id": "104e6d3699e",
  "name": "snap1",
  "state": 1,
  "image_snap_name": ".group.3_103a6510642_104e6d3699e",
  "images": [
    {
      "pool_id": 3,
      "snap_id": 3,
      "image_name": "img1"
    },
    {
      "pool_id": 3,
      "snap_id": 4,
      "image_name": "img2"
    },
    {
      "pool_id": 3,
      "snap_id": 5,
      "image_name": "img3"
    }
  ]
}
{
  "id": "1050aa0ea3dc",
  "name": "snap2",
  "state": 1,
  "image_snap_name": ".group.3_103a6510642_1050aa0ea3dc",
  "images": [
    {
      "pool_id": 3,
      "snap_id": 6,
      "image_name": "img1"
    },
    {
      "pool_id": 3,
      "snap_id": 7,
      "image_name": "img2"
    },
    {
      "pool_id": 3,
      "snap_id": 8,
      "image_name": "img3"
    }
  ]
}
{
  "id": "1052641a00bb",
  "name": "snap3",
  "state": 1,
  "image_snap_name": ".group.3_103a6510642_1052641a00bb",
  "images": [
    {
      "pool_id": 3,
      "snap_id": 9,
      "image_name": "img1"
    },
    {
      "pool_id": 3,
      "snap_id": 10,
      "image_name": "img2"
    },
    {
      "pool_id": 3,
      "snap_id": 11,
      "image_name": "img3"
    }
  ]
```

ajarr commented:
> @ajarr Is it expected to have `"snap_name": ".group.3_103e77dd2be9_104ccb3288f3"` starting with `.`?

@sunilangadi2 please see https://github.com/ceph/ceph/pull/57797#issuecomment-2177328767

Yes, the snapshots of images that belong to a group snapshot have names starting with .**group**. We (developers) refer to them as "dotgroup" snapshots.

idryomov commented on src/tools/rbd/action/Group.cc:
```diff
@@ -704,6 +706,111 @@ int execute_group_snap_list(const po::variables_map &vm,
+
+   int state = group_snap.state;
+   std::string state_string;
+   if (RBD_GROUP_SNAP_STATE_INCOMPLETE == state) {
```
... and since this is a new command, and we disrupting the existing command by adding the ID column anyway, I'd consider changing "ok" to "complete" in both commands to actually reflect the name of the state.

idryomov reviewed (commented):
Commenting on the output, will have a few nits on the implementation later.

nbalacha commented on src/librbd/api/Group.cc:
```diff
@@ -1211,22 +1201,119 @@ int Group<I>::snap_rename(librados::IoCtx& group_ioctx, const char *group_name,
+    image_snaps.reserve(cls_group_snap.snaps.size());
+    for (const auto &i : cls_group_snap.snaps) {
+      std::string image_name;
+      r = cls_client::dir_get_name(&group_ioctx, RBD_DIRECTORY,
```
This may not work if the image is in a different pool.

nbalacha commented on src/librbd/api/Group.cc:
```diff
@@ -1211,22 +1201,119 @@ int Group<I>::snap_rename(librados::IoCtx& group_ioctx, const char *group_name,
+
+  for (const auto &i : cls_group_snap->snaps) {
+    std::string image_name;
+    r = cls_client::dir_get_name(&group_ioctx, RBD_DIRECTORY,
```
The image may be in a different group.

ajarr commented:
TODOS:
- write tests
- cleanup python docstrings

ajarr commented:
jenkins test api

ajarr commented:
jenkins test make check

idryomov commented on src/test/librbd/test_Groups.cc:
```diff
@@ -480,3 +480,241 @@ TEST_F(TestGroup, add_snapshotPP)
+  const char *gp_snap_name2 = "snap_snapshot2";
+  ASSERT_EQ(0, rbd_group_image_remove(ioctx, gp_name, ioctx, image_name1));
+  ASSERT_EQ(0, rbd_group_image_add(ioctx, gp_name, ioctx2, image_name2));
+  ASSERT_EQ(0, rbd_group_snap_create(ioctx, gp_name, gp_snap_name2));
```
For completeness, I'd take one more snapshot here -- with both images in the group.

idryomov commented on src/test/librbd/test_Groups.cc:
```diff
@@ -480,3 +480,241 @@ TEST_F(TestGroup, add_snapshotPP)
+  uint64_t features;
+  int order =0;
+  ASSERT_TRUE(get_features(&features));
+  ASSERT_EQ(0, rbd_create2(ioctx, image_name1, m_image_size, features,
```
Can an image that is automatically created be used instead?

nbalacha commented on src/tools/rbd/action/Group.cc:
```diff
@@ -756,6 +758,111 @@ int execute_group_snap_list(const po::variables_map &vm,
+
+    if (f) {
+      f->open_object_section("image");
+      f->dump_string("pool_name", pool_name);
```
This will attempt to print pool_name even if the pool_reverse_lookup() call fails.

nbalacha commented on src/test/librbd/test_Groups.cc:
```diff
@@ -480,3 +480,241 @@ TEST_F(TestGroup, add_snapshotPP)
+      FAIL() << "Unexpected group snap: " << gp_snaps[i].name;
+    }
+  }
+  ASSERT_EQ(0, rbd_group_snap_remove(ioctx, gp_name, gp_snap_name1));
```
Doesn't gp_snap_name0 need to be removed?

nbalacha commented on src/test/librbd/test_Groups.cc:
```diff
@@ -480,3 +480,241 @@ TEST_F(TestGroup, add_snapshotPP)
+      FAIL() << "Unexpected group snap: " << i.name;
+    }
+  }
+
```
It would be nice to list the snaps again after removing one or more snaps to make sure they have been removed.

ajarr commented on src/librbd/librbd.cc:
```diff
@@ -7230,6 +7280,48 @@ extern "C" int rbd_group_snap_rename(rados_ioctx_t group_p,
+  group_snap->name = NULL;
+  group_snap->image_snap_name = NULL;
+  group_snap->image_snaps = NULL;
+
```
Thanks for the explanation!

ajarr commented on src/librbd/librbd.cc:
```diff
@@ -7293,6 +7385,57 @@ extern "C" int rbd_group_snap_list_cleanup(rbd_group_snap_info_t *snaps,
+
+  if (r == -ENOENT) {
+    *snaps_size = 0;
+    return 0;
```
Good point!

ajarr commented on src/tools/rbd/action/Group.cc:
```diff
@@ -705,10 +705,9 @@ int execute_group_snap_list(const po::variables_map &vm,
+  r = rbd.group_snap_list2(io_ctx, group_name.c_str(), &snaps);
 
   if (r == -ENOENT) {
     r = 0;
```
Good catch!

ajarr commented on src/include/rbd/librbd.h:
```diff
@@ -249,11 +249,27 @@ typedef enum {
+  char *name;
+  char *image_snap_name;
+  rbd_group_snap_state_t state;
+  rbd_snap_namespace_type_t unused;
```
Added comment here https://github.com/ceph/ceph/pull/53793#discussion_r1665843612

ajarr commented on PendingReleaseNotes:
```diff
@@ -234,6 +234,11 @@ CephFS: Disallow delegating preallocated inode ranges to clients. Config
+  show information about a group and a group snapshot respectively.
+* RBD: `rbd group snap ls` output now includes the group snap IDs. The 'state'
+  of a group snapshot that was shown as 'okay' is now shown as 'complete',
+  which is more descriptive.
```
I've deliberately not fixed this to avoid rebase and make it easy to review the changes between pushes. I will make this change once the changes look good.

ajarr commented on src/include/rbd/librbd.hpp:
```diff
@@ -161,11 +161,26 @@ namespace librbd {
+    std::string name;
+    std::string image_snap_name;
+    group_snap_state_t state;
+    snap_namespace_type_t unused;
```
Added comment here https://github.com/ceph/ceph/pull/53793#discussion_r1671334259

ajarr commented on src/pybind/rbd/rbd.pyx:
```diff
@@ -2775,6 +2775,69 @@ cdef class Group(object):
+                              group_errno_to_exception)
+            image_snaps = []
+            for i in range(group_snap.image_snaps_count):
+                image_snap = group_snap.image_snaps[i]
```
Using `&` and not using it works for me. It looks Cython seems to take care of it.

ajarr commented on src/pybind/rbd/rbd.pyx:
```diff
@@ -5928,47 +5991,72 @@ cdef class GroupSnapIterator(object):
     def __iter__(self):
-        for i in range(self.num_snaps):
+        for i in range(self.num_group_snaps):
+            group_snap = self.group_snaps[i]
```
Using & and not using it works for me. It looks Cython seems to take care of it.

ajarr commented:
See https://github.com/ceph/ceph/compare/393d60c65323f87f7cc9..56e79f98b901f6493f20ab271a2304f835940001 to view the changes made. I've addressed all the comments except the one in release notes to avoid rebase and help with reviewing only the changes.

ajarr commented on src/librbd/api/Group.cc:
```diff
@@ -1226,24 +1216,118 @@ int Group<I>::snap_rename(librados::IoCtx& group_ioctx, const char *group_name,
   return 0;
 }
 
+int convert_cls_rbd_GroupSnapshot_to_librbd_group_snap_info2_t(
```
Suggestions welcome for a less kludgy function name

idryomov replied:
I'd go with `GroupSnapshot_to_group_snap_info2()`.

idryomov commented on src/librbd/api/Group.cc:
```diff
@@ -1226,24 +1216,118 @@ int Group<I>::snap_rename(librados::IoCtx& group_ioctx, const char *group_name,
+      return r;
+    }
+
+    image_snaps.emplace_back(image_name, snap.pool, snap.snap_id);
```
This doesn't build, at least with clang:
```
[2454/2755] Building CXX object src/librbd/CMakeFiles/rbd_internal.dir/api/Group.cc.o
FAILED: src/librbd/CMakeFiles/rbd_internal.dir/api/Group.cc.o 
/usr/bin/ccache /usr/bin/clang++-14 -DBOOST_ASIO_DISABLE_THREAD_KEYWORD_EXTENSION -DBOOST_ASIO_HAS_IO_URING -DBOOST_ASIO_NO_TS_EXECUTORS -DBOOST_DATE_TIME_DYN_LINK -DBOOST_DATE_TIME_NO_LIB -DBOOST_IOSTREAMS_DYN_LINK -DBOOST_IOSTREAMS_NO_LIB -DBOOST_PROGRAM_OPTIONS_DYN_LINK -DBOOST_PROGRAM_OPTIONS_NO_LIB -DBOOST_RANDOM_DYN_LINK -DBOOST_RANDOM_NO_LIB -DBOOST_SYSTEM_DYN_LINK -DBOOST_SYSTEM_NO_LIB -DBOOST_THREAD_DYN_LINK -DBOOST_THREAD_NO_LIB -DHAVE_CONFIG_H -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -D_REENTRANT -D_THREAD_SAFE -D__CEPH__ -D__STDC_FORMAT_MACROS -D__linux__ -I/home/jenkins-build/build/workspace/ceph-pull-requests/build/src/include -I/home/jenkins-build/build/workspace/ceph-pull-requests/src -isystem /opt/ceph/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/build/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/src/jaegertracing/opentelemetry-cpp/api/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/src/jaegertracing/opentelemetry-cpp/exporters/jaeger/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/src/jaegertracing/opentelemetry-cpp/ext/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/src/jaegertracing/opentelemetry-cpp/sdk/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/src/xxHash -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/src/fmt/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/build/src/pmdk/src/include -g -Werror -fPIC -fno-builtin-malloc -fno-builtin-calloc -fno-builtin-realloc -fno-builtin-free -Wall -fno-strict-aliasing -fsigned-char -Wtype-limits -Wignored-qualifiers -Wpointer-arith -Werror=format-security -Winit-self -Wno-unknown-pragmas -Wnon-virtual-dtor -Wno-ignored-qualifiers -ftemplate-depth-1024 -Wpessimizing-move -Wredundant-move -Wno-inconsistent-missing-override -Wno-mismatched-tags -Wno-unused-private-field -Wno-address-of-packed-member -Wno-unused-function -Wno-unused-local-typedef -Wno-varargs -Wno-gnu-designator -Wno-missing-braces -Wno-parentheses -Wno-deprecated-register -DCEPH_DEBUG_MUTEX -D_GLIBCXX_ASSERTIONS -fdiagnostics-color=auto -std=c++20 -MD -MT src/librbd/CMakeFiles/rbd_internal.dir/api/Group.cc.o -MF src/librbd/CMakeFiles/rbd_internal.dir/api/Group.cc.o.d -o src/librbd/CMakeFiles/rbd_internal.dir/api/Group.cc.o -c /home/jenkins-build/build/workspace/ceph-pull-requests/src/librbd/api/Group.cc
In file included from /home/jenkins-build/build/workspace/ceph-pull-requests/src/librbd/api/Group.cc:4:
In file included from /home/jenkins-build/build/workspace/ceph-pull-requests/src/common/Cond.h:19:
In file included from /home/jenkins-build/build/workspace/ceph-pull-requests/src/common/Clock.h:18:
In file included from /home/jenkins-build/build/workspace/ceph-pull-requests/src/include/utime.h:28:
In file included from /home/jenkins-build/build/workspace/ceph-pull-requests/src/include/types.h:21:
In file included from /home/jenkins-build/build/workspace/ceph-pull-requests/src/include/uuid.h:9:
In file included from /home/jenkins-build/build/workspace/ceph-pull-requests/src/include/encoding.h:18:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/set:60:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/stl_tree.h:67:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/ext/alloc_traits.h:34:
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/alloc_traits.h:518:4: error: no matching function for call to 'construct_at'
          std::construct_at(__p, std::forward<_Args>(__args)...);
          ^~~~~~~~~~~~~~~~~
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/vector.tcc:117:21: note: in instantiation of function template specialization 'std::allocator_traits<std::allocator<librbd::group_image_snap_info_t>>::construct<librbd::group_image_snap_info_t, std::basic_string<char> &, const long &, const snapid_t &>' requested here
            _Alloc_traits::construct(this->_M_impl, this->_M_impl._M_finish,
                           ^
/home/jenkins-build/build/workspace/ceph-pull-requests/src/librbd/api/Group.cc:1242:17: note: in instantiation of function template specialization 'std::vector<librbd::group_image_snap_info_t>::emplace_back<std::basic_string<char> &, const long &, const snapid_t &>' requested here
    image_snaps.emplace_back(image_name, snap.pool, snap.snap_id);
                ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/stl_construct.h:94:5: note: candidate template ignored: substitution failure [with _Tp = librbd::group_image_snap_info_t, _Args = <std::basic_string<char> &, const long &, const snapid_t &>]: no matching constructor for initialization of 'librbd::group_image_snap_info_t'
    construct_at(_Tp* __location, _Args&&... __args)
    ^
1 error generated.
```
I think the issue is that `ImageSnapshotSpec`'s `snap_id` and `group_image_snap_info_t`'s `snap_id` aren't the same type (`snapid_t` which is a struct vs `uint64_t`) and the implicit conversion may not be picked up in this context.

ajarr replied:
Let's see if explicit conversion to uint64_t fixes it

ajarr replied:
Still fails

idryomov replied:
Sigh, Clang 14 just doesn't support this C++20 feature.  The following trivial example doesn't compile for me either:
```
#include <vector>

struct foo {
    int i;
};

int main() {
    std::vector<foo> v;
    v.emplace_back(1);
}
```
According to https://en.cppreference.com/w/cpp/compiler_support/20, "Parenthesized initialization of aggregates" is supported only starting with Clang 16.

ajarr replied:
Switched back to using `push_back()`

idryomov commented on src/librbd/api/Group.cc:
```diff
@@ -1226,24 +1216,118 @@ int Group<I>::snap_rename(librados::IoCtx& group_ioctx, const char *group_name,
 
+int convert_cls_rbd_GroupSnapshot_to_librbd_group_snap_info2_t(
+    librados::IoCtx& group_ioctx, const std::string& group_id,
+    const cls::rbd::GroupSnapshot* cls_group_snap,
```
Observation-only, never-NULL parameters should be passed for const reference:
```suggestion
    const cls::rbd::GroupSnapshot& cls_group_snap,
```

nbalacha commented on src/librbd/librbd.cc:
```diff
@@ -263,12 +263,31 @@ void group_info_cpp_to_c(const librbd::group_info_t &cpp_info,
+  c_info->image_snap_name = strdup(cpp_info.image_snap_name.c_str());
+  c_info->state = cpp_info.state;
+  c_info->image_snaps_count = cpp_info.image_snaps.size();
+  c_info->image_snaps = static_cast<rbd_group_image_snap_info_t*>(calloc(
```
Check for NULL return (calloc failure).

idryomov replied:
The rest of the `src/librbd/librbd.cc` doesn't check for memory allocation errors in these helpers, most likely because it's very hard to get any `malloc()`-like function to fail on Linux.

If we want to be serious to get serious about it, returns from `strdup()`, etc should be checked as well, and that should be done consistently.  It's a topic for another tracker/PR.

nbalacha replied:
Yes, all the memory allocation functions should be checked and the functions should return an error if any of them fail. I agree, it can be a separate PR.

nbalacha commented on src/librbd/librbd.cc:
```diff
@@ -1436,11 +1455,33 @@ namespace librbd {
-    int r = librbd::api::Group<>::snap_list(group_ioctx, group_name, snaps);
+    std::vector<group_snap_info2_t> snaps2;
+    int r = librbd::api::Group<>::snap_list(group_ioctx, group_name, &snaps2);
+    for (const auto& snap : snaps2) {
```
Will snaps2 always be empty in case the function returns an error?

nbalacha commented on src/librbd/librbd.cc:
```diff
@@ -7293,6 +7362,43 @@ extern "C" int rbd_group_snap_list_cleanup(rbd_group_snap_info_t *snaps,
+                                    size_t *snaps_size)
+{
+  // FIPS zeroization audit 20191117: this memset is not security related.
+  memset(snaps, 0, sizeof(*snaps) * *snaps_size);
```
This can be moved to line 7386.

idryomov replied:
I think this memset should be removed altogether:

> [...] (and I'm not convinced that memset is needed either -- see the comment on "random junk").

nbalacha commented on src/librbd/librbd.cc:
```diff
@@ -7293,6 +7362,43 @@ extern "C" int rbd_group_snap_list_cleanup(rbd_group_snap_info_t *snaps,
+}
+
+extern "C" void rbd_group_snap_list2_cleanup(rbd_group_snap_info2_t *snaps,
+                                             size_t len) {
```
Nit. : this is not really a len. Perhaps count or num?

ajarr commented on qa/workunits/rbd/rbd_groups.sh:
```diff
@@ -182,14 +182,65 @@ check_snapshot_not_in_group()
+        jq -r '.id')
+    [[ ! -z "$snap_id_in_info" ]] || return 1
+
+    local snap_list=$(rbd group snap ls $group_name --format=json | jq -c '.[]')
```
Nice suggestion!

ajarr commented:
Please see https://github.com/ceph/ceph/compare/56e79f98b901f64..f0c89a8c1622833746f030e3771f2f238f52c9f7 for the new changes.

nbalacha commented on src/test/librbd/test_Groups.cc:
```diff
@@ -480,3 +480,252 @@ TEST_F(TestGroup, add_snapshotPP)
+
+  for (const auto& gp_snap_name : gp_snap_names) {
+    ASSERT_EQ(0, m_rbd.group_snap_remove(m_ioctx, gp_name, gp_snap_name));
+  }
```
It would be nice to check that the listing returns 0 snaps now.

idryomov commented on src/test/pybind/test_rbd.py:
```diff
@@ -2882,6 +2882,38 @@ def test_group_image_remove(self):
+        img_snap_keys = sorted(['pool_id', 'snap_id', 'image_name'])
+        assert sorted(snap_info_dict.keys()) == (
+            sorted(['id', 'name', 'state', 'image_snap_name', 'image_snaps']))
+        assert snap_info_dict['name'] == snap_name
```
Let's add an assert for state -- it should be `RBD_GROUP_SNAP_STATE_COMPLETE`.

idryomov commented on src/test/pybind/test_rbd.py:
```diff
@@ -2919,18 +2951,37 @@ def test_group_snap_flags(self):
+            assert sorted(gp_snap.keys()) == gp_snap_keys
+            gp_snap_names.append(gp_snap['name'])
+            image_names = []
+            for i in gp_snap['image_snaps']:
```
```suggestion
            for img_snap in gp_snap['image_snaps']:
```

idryomov commented on src/test/pybind/test_rbd.py:
```diff
@@ -2919,18 +2951,37 @@ def test_group_snap_flags(self):
+            image_names = []
+            for i in gp_snap['image_snaps']:
+                assert sorted(i.keys()) == img_snap_keys
+                assert i['pool_id'] == pool_id
```
I'd drop this assert, leaving this test entirely to names -- it's pretty redundant as it is.

idryomov commented on src/test/pybind/test_rbd.py:
```diff
@@ -2919,18 +2951,37 @@ def test_group_snap_flags(self):
+                assert i['pool_id'] == pool_id
+                image_names.append(i['image_name'])
+            assert sorted(image_names) == self.image_names
+        gp_snap_names.sort()
```
How do you choose between `sorted()` (new list) and `sort()` (in-place)?  It doesn't seem to be consistent to me -- is there a (non-obvious) guideline to follow?

ajarr replied:
`sort()` cannot be used for iterables such as dictionary keys. For lists, `sort()` is preferred if it's okay to modify in place instead of creating a new list. I've switched to using `sorted()` throughout.

ajarr replied:
https://stackoverflow.com/a/22442750

idryomov commented on src/test/pybind/test_rbd.py:
```diff
@@ -2882,6 +2882,38 @@ def test_group_image_remove(self):
+
+        assert sorted(image_names) == self.image_names
+
+        self.group.remove_snap(snap_name)
```
Missing a negative test for a non-existing snapshot, expecting an exception.  One passing `snap_name` could be added here -- otherwise the cleanup could be omitted entirely.

idryomov commented on src/test/pybind/test_rbd.py:
```diff
@@ -2882,6 +2882,38 @@ def test_group_image_remove(self):
+
+        self.group.create_snap(snap_name)
+
+        snap_info_dict = self.group.get_snap_info(snap_name)
```
> Missing a negative test for a non-existing snapshot, expecting an exception.

... and one passing an empty string could be added here, where there is some existing snapshot.

idryomov reviewed (commented):
I haven't looked at `src/test/librbd/test_Groups.cc` this time -- will do on the next iteration.

nbalacha commented:
I think you also need to modify `doc/man/8/rbd.rst`

ajarr commented on src/librbd/librbd.cc:
```diff
@@ -1436,11 +1455,38 @@ namespace librbd {
+    if (r < 0) {
+      tracepoint(librbd, group_snap_list_exit, r);
+      return r;
+    }
```
Will do this. Nice suggestion!

ajarr commented on src/test/pybind/test_rbd.py:
```diff
@@ -2882,6 +2882,38 @@ def test_group_image_remove(self):
+            sorted(['id', 'name', 'state', 'image_snap_name', 'image_snaps']))
+        assert snap_info_dict['name'] == snap_name
+        for image_snap in snap_info_dict['image_snaps']:
+            assert sorted(image_snap.keys()) == img_snap_keys
```
This is to check that the dictionary has exactly those three keys and nothing more. I can remove it

ajarr commented:
Please see https://github.com/ceph/ceph/compare/87c73d4232c91c1e..062dd70c958f1b3541b990a7aa798a28a2c7df86

idryomov commented on qa/workunits/rbd/rbd_groups.sh:
```diff
@@ -182,12 +182,42 @@ check_snapshot_not_in_group()
+    local actual_snap_name=$(jq -r ".name" <<< "$snap_info")
+    test "$actual_snap_name" = "$snap_name" || return 1
+
+    [[ -z "$image_name" ]] && return 0
```
> In the test you could create an additional snapshot right after the group is created to exercise this scenario.

This appears to have been missed.

ajarr replied:
TODO

idryomov commented on src/test/librbd/test_Groups.cc:
```diff
@@ -480,3 +480,255 @@ TEST_F(TestGroup, add_snapshotPP)
+  rados_ioctx_t ioctx2;
+  rados_ioctx_create(_cluster, pool_name2.c_str(), &ioctx2);
+
+  const char *gp_name = "snap_group";
```
Make these names unique at least across the file, see commit 4c69370f6b0e4c04e77fde67a9c199af6ea13649.

idryomov commented on src/test/librbd/test_Groups.cc:
```diff
@@ -480,3 +480,255 @@ TEST_F(TestGroup, add_snapshotPP)
+  ASSERT_EQ(0, rbd_group_create(ioctx2, gp_name));
+  ASSERT_EQ(0, rbd_group_image_add(ioctx2, gp_name, ioctx, image_name));
+
+
```
duplicate blank line

idryomov commented on src/test/librbd/test_Groups.cc:
```diff
@@ -480,3 +480,255 @@ TEST_F(TestGroup, add_snapshotPP)
+  rados_ioctx_create(_cluster, pool_name2.c_str(), &ioctx2);
+
+  const char *gp_name = "snap_group";
+  const char *image_name = m_image_name.c_str();
```
This variable is redundant.

idryomov commented on src/test/librbd/test_Groups.cc:
```diff
@@ -480,3 +480,255 @@ TEST_F(TestGroup, add_snapshotPP)
+  ASSERT_EQ(0, _rados.ioctx_create(pool_name2.c_str(), ioctx2));
+
+  const char *gp_name = "snap_group";
+  const char *image_name = m_image_name.c_str();
```
ditto

idryomov commented on src/test/librbd/test_Groups.cc:
```diff
@@ -480,3 +480,255 @@ TEST_F(TestGroup, add_snapshotPP)
+  const char *image_name1 = m_image_name.c_str();
+  const char *image_name2 = "snap_list_image2";
+  uint64_t features;
+  int order =0;
```
```suggestion
  int order = 0;
```

idryomov commented on src/test/librbd/test_Groups.cc:
```diff
@@ -480,3 +480,255 @@ TEST_F(TestGroup, add_snapshotPP)
+  rados_ioctx_t ioctx2;
+  rados_ioctx_create(_cluster, pool_name2.c_str(), &ioctx2);
+
+  const char *image_name1 = m_image_name.c_str();
+  const char *image_name2 = "snap_list_image2";
```
`image_name1` is redundant, no need to introduce aliases.  Notice how the alias is also inconsistent with ioctx naming: we have `ioctx` and `ioctx2` but `image_name1` and `image_name2`.

For `image_name2`, it would be better to use `get_temp_image_name()`.

idryomov commented on src/test/librbd/test_Groups.cc:
```diff
@@ -480,3 +480,255 @@ TEST_F(TestGroup, add_snapshotPP)
+  ASSERT_EQ(0, rbd_group_image_remove(ioctx, gp_name, ioctx, image_name1));
+  ASSERT_EQ(0, rbd_group_snap_create(ioctx, gp_name, gp_snap_names[3]));
+
+  num_snaps= 10U;
```
```suggestion
  num_snaps = 10U;
```

idryomov commented on src/test/librbd/test_Groups.cc:
```diff
@@ -480,3 +480,255 @@ TEST_F(TestGroup, add_snapshotPP)
+  ASSERT_EQ(0, rbd_group_snap_create(ioctx, gp_name, gp_snap_names[3]));
+
+  num_snaps= 10U;
+  ASSERT_EQ(0, rbd_group_snap_list2(ioctx, gp_name, gp_snaps, &num_snaps));
```
Add an additional case above, setting `num_snaps` to 3 and expecting ERANGE and `num_snaps` getting reset to 4 by the API.

idryomov commented on src/test/librbd/test_Groups.cc:
```diff
@@ -480,3 +480,255 @@ TEST_F(TestGroup, add_snapshotPP)
+	if (!strcmp(gp_snaps[i].image_snaps[j].image_name, image_name1)) {
+	  ASSERT_EQ(rados_ioctx_get_id(ioctx),
+                    gp_snaps[i].image_snaps[j].pool_id);
+	} else if (!strcmp(gp_snaps[i].image_snaps[j].image_name, image_name2))
+        {
```
```suggestion
	} else if (!strcmp(gp_snaps[i].image_snaps[j].image_name, image_name2)) {
```

idryomov commented on src/test/librbd/test_Groups.cc:
```diff
@@ -480,3 +480,255 @@ TEST_F(TestGroup, add_snapshotPP)
+  librados::IoCtx ioctx2;
+  ASSERT_EQ(0, _rados.ioctx_create(pool_name2.c_str(), ioctx2));
+
+  const char *image_name1=m_image_name.c_str();
+
+  const char *image_name2="snap_list2PP_image2";
```
ditto
```suggestion
  std::string image_name2 = get_temp_image_name();
```

idryomov commented on src/test/librbd/test_Groups.cc:
```diff
@@ -480,3 +480,255 @@ TEST_F(TestGroup, add_snapshotPP)
+  }
+  std::vector<librbd::group_snap_info2_t> gp_snaps2;
+  ASSERT_EQ(0, m_rbd.group_snap_list2(m_ioctx, gp_name, &gp_snaps2));
+  ASSERT_EQ(0U, gp_snaps2.size());
```
Add a corresponding "listing is now empty" case to the C API test for consistency.

idryomov commented:
> I think you also need to modify `doc/man/8/rbd.rst`

This is still pending -- bumping since it wasn't a comment on the diff and therefore may get missed.

ajarr commented on src/test/pybind/test_rbd.py:
```diff
@@ -2882,6 +2882,36 @@ def test_group_image_remove(self):
+        self.group.create_snap(snap_name)
+        snap_info_dict = self.group.get_snap_info(snap_name)
+        image_names = []
+        img_snap_keys = sorted(['pool_id', 'snap_id', 'image_name'])
```
TODO

ajarr commented:
> > I think you also need to modify `doc/man/8/rbd.rst`
> 
> This is still pending -- bumping since it wasn't a comment on the diff and therefore may get missed.

Done

ajarr commented:
Please see https://github.com/ceph/ceph/compare/062dd70c958f1b3541b990a7aa798a28a2c7df86..b146d3a63236339460eb1bf641226f5a730350c9

idryomov commented on qa/workunits/rbd/rbd_groups.sh:
```diff
@@ -217,23 +248,25 @@ echo "PASSED"
+snaps=("group_snap1" "group_snap2" "group_snap3" "group_snap4")
 create_image $image
 create_group $group
+create_snapshot $group ${snaps[0]}
```
I'd pass an empty image name explicitly:
```suggestion
create_snapshot $group ${snaps[0]} ""
```

idryomov commented on qa/workunits/rbd/rbd_groups.sh:
```diff
@@ -217,23 +248,25 @@ echo "PASSED"
-remove_snapshot $group $sec_snap
-check_snapshot_not_in_group $group $sec_snap
+create_snapshot $group ${snaps[1]}
+check_snapshot_in_group $group ${snaps[1]}
```
This call seems redundant as `check_snapshot_info()` is called immediately after and would fail if the snapshot isn't there.

idryomov commented on src/test/librbd/test_Groups.cc:
```diff
@@ -480,3 +480,262 @@ TEST_F(TestGroup, add_snapshotPP)
+  }
+  rbd_group_snap_list2_cleanup(gp_snaps, num_snaps);
+  free(gp_snaps);
+  ASSERT_EQ(0, rbd_group_snap_list2(ioctx, gp_name, gp_snaps, &num_snaps));
```
Passing NULL would be better since `gp_snaps` just got freed:
```suggestion
  ASSERT_EQ(0, rbd_group_snap_list2(ioctx, gp_name, NULL, &num_snaps));
```

idryomov reviewed (commented):
Mostly `c_str()` redux ;)

nbalacha commented on src/librbd/api/Group.cc:
```diff
@@ -53,26 +53,16 @@ snap_t get_group_snap_id(I* ictx,
 }
 
-int group_snap_list(librados::IoCtx& group_ioctx, const char *group_name,
+int group_snap_list(librados::IoCtx& group_ioctx, const std::string& group_id,
```
This change seems unrelated to this PR. Why is it required?

idryomov replied:
This is to avoid calling `dir_get_id` CLS method twice -- currently all but one caller of `group_snap_list()` do it.

ajarr replied:
Correct

nbalacha commented:
Looks good to me pending addressing Ilya's comments.

ajarr commented on src/test/librbd/test_Groups.cc:
```diff
@@ -480,3 +480,262 @@ TEST_F(TestGroup, add_snapshotPP)
+  ASSERT_EQ(0, rados_pool_create(_cluster, pool_name2.c_str()));
+
+  rados_ioctx_t ioctx;
+  rados_ioctx_create(_cluster, _pool_name.c_str(), &ioctx);
```
TODO

ajarr commented:
https://github.com/ceph/ceph/compare/b146d3a..d5bf5888d9941b68885fe0ebdd0c4dc14be7b3fc

idryomov commented:
I think this is ready for a rebase to address the final outstanding comment!

It would be nice post a new set of outputs as well -- the set in the comment that is linked from the description is outdated (it still has `state: ok`) and it also doesn't demonstrate images in multiple pools.

idryomov commented on src/tools/rbd/action/Group.cc:
```diff
@@ -721,38 +727,127 @@ int execute_group_snap_list(const po::variables_map &vm,
   } else {
+    t.define_column("ID", TextTable::LEFT, TextTable::LEFT);
     t.define_column("NAME", TextTable::LEFT, TextTable::LEFT);
     t.define_column("STATUS", TextTable::LEFT, TextTable::RIGHT);
```
```
$ rbd group snap ls data/gp
ID            NAME   STATUS  
104eb2666faf  snap1  complete
```
Since we have changed what is displayed in this column, let's take this opportunity to also change its name to be consistent with the name of the corresponding field:
```suggestion
    t.define_column("STATE", TextTable::LEFT, TextTable::RIGHT);
```

ajarr replied:
Done.  TODO Need to update ReleaseNotes too.

idryomov commented on PendingReleaseNotes:
```diff
@@ -6,6 +6,11 @@
   methods treat "naive" `datetime` objects as local times.
+* RBD: `rbd group info` and `rbd group snap info` commands are introduced to
+  show information about a group and a group snapshot respectively.
+* RBD: `rbd group snap ls` output now includes the group snap IDs. The 'state'
+  of a group snapshot that was shown as 'okay' is now shown as 'complete',
```
```suggestion
* RBD: `rbd group snap ls` output now includes the group snap IDs. The state
  of a group snapshot that was shown as 'ok' is now shown as 'complete',
```

idryomov commented on src/tools/rbd/action/Group.cc:
```diff
@@ -721,38 +727,127 @@ int execute_group_snap_list(const po::variables_map &vm,
+  }
+
+  auto pname_itr = pool_names.begin();
+  for (const auto& image_snap : group_snap.image_snaps) {
```
```
$ rbd group snap info data/gp@snap4
...
	images:
		rbd/img4 (snap id: 339)
		data/img1 (snap id: 12)
		data/img2 (snap id: 13)
		data/img3 (snap id: 14)
```
Since at its core this is an image listing, I think it would be nice to be consistent with the output of `rbd group image ls` command.  I see two differences:
- `rbd group image ls` sorts images based on pool ID and image name: https://github.com/ceph/ceph/blob/103cd8e78bcfe7f69647013187c053c9ccb76685/src/tools/rbd/action/Group.cc#L487-L495
- `rbd group image ls` doesn't fail if the pool name can't be produced and it's actually important.  I'll expand in a separate comment.

Do you think we could stick a similar `std::sort()` invocation here?

idryomov commented on src/tools/rbd/action/Group.cc:
```diff
@@ -721,38 +727,127 @@ int execute_group_snap_list(const po::variables_map &vm,
     if (r < 0) {
+      std::cerr << "rbd: error looking up pool name for pool_id=" << pool_id
+                << ": " << cpp_strerror(r) << std::endl;
       return r;
```
`rbd group image ls` produces a `"<missing image pool $POOL_ID>"` string instead of failing in this case and for a good reason: if a user deletes a pool with one of the group's images, we still want to be able to get information about that group or its snapshots.  We should do the same here.

It also occurs to me that it would be better to make `pool_names` a map from pool ID to pool name rather than a vector of pool names.  That way we wouldn't be storing duplicate strings (a group is likely to have all of its images come from a single pool) and it would be immune to sorting.

idryomov replied:
> it would be better to make `pool_names` a map from pool ID to pool name rather than a vector of pool names

... or we could do exactly what `execute_list_images()` does and perform the lookup inside the loop, eliminating a separate data structure entirely.

ajarr replied:
Now does lookup inside the loop. Uses a map to not redo look up of pool name.

idryomov replied:
See https://github.com/ceph/ceph/pull/57797#discussion_r1681716952.

ajarr replied:
I will create a tracker ticket for the issues with the `group` set of CLI commands when the pool of a group's image is removed.

ajarr replied:
https://tracker.ceph.com/issues/67021

idryomov replied:
Thanks for looking into this!

idryomov commented on src/tools/rbd/action/Group.cc:
```diff
@@ -756,6 +754,112 @@ int execute_group_snap_list(const po::variables_map &vm,
+    }
+  );
+
+  std::map<int64_t, std::string> pool_names;
```
Honestly, I'd drop the map and just call `pool_reverse_lookup()` on each iteration.  Caching pool names in a scenario like this seems a bit silly to me -- these lookups are in-memory and cheap, no round-trip to the cluster or anything like that is involved.

idryomov replied:
> no round-trip to the cluster or anything like that is involved.

Correction: round-trips to the monitor are involved in the pool isn't found, to "confirm" the prospective `ENOENT` error.  But you aren't caching negative lookups here and the case of a pool getting removed underneath a group isn't something that we should be worrying about optimizing anyway, so the point I made in the previous comment stands.

ajarr replied:
Got it. Thanks for the explanation!

idryomov commented on src/tools/rbd/action/Group.cc:
```diff
@@ -756,6 +754,105 @@ int execute_group_snap_list(const po::variables_map &vm,
+  for (const auto& image_snap : group_snap.image_snaps) {
+    std::string pool_name;
+    r = rados.pool_reverse_lookup(image_snap.pool_id, &pool_name);
+    if (r < 0) {
+      pool_name = "<missing image pool " + stringify(image_snap.pool_id) + ">";
+    }
```
Theoretically, other (transient) errors can be returned as well:
```suggestion
    if (r == -ENOENT) {
      pool_name = "<missing image pool " + stringify(image_snap.pool_id) + ">";
    } else if (r < 0) {
      std::cerr << "rbd: error looking up pool name for pool_id=" << pool_id
                << ": " << cpp_strerror(r) << std::endl;
      return r;
    }
```

idryomov commented on src/tools/rbd/action/Group.cc:
```diff
@@ -756,6 +754,105 @@ int execute_group_snap_list(const po::variables_map &vm,
+  }
+
+  std::sort(group_snap.image_snaps.begin(), group_snap.image_snaps.end(),
+    [](const librbd::group_image_snap_info_t &lhs,
+       const librbd::group_image_snap_info_t &rhs) {
```
```suggestion
    [](const librbd::group_image_snap_info_t& lhs,
       const librbd::group_image_snap_info_t& rhs) {
```

ajarr commented:
@idryomov as you mentioned earlier,  previous implementation of `group snap ls` command that displayed just the snap name and state would work when the pool of a group's image is removed from underneath.  The new implementation of `group snap ls` introduced in this PR will fail in such a scenario (See https://tracker.ceph.com/issues/67021). I guess we'll fix this properly when working on https://tracker.ceph.com/issues/67021 ?

idryomov commented:
No related failures:

https://pulpito.ceph.com/dis-2024-07-18_22:11:30-rbd-wip-dis-testing-distro-default-smithi/
https://pulpito.ceph.com/dis-2024-07-19_12:06:54-rbd-wip-dis-testing-distro-default-smithi/
https://pulpito.ceph.com/dis-2024-07-19_20:40:00-rbd-wip-dis-testing-distro-default-smithi/

(Jobs where `TestLibRBD.TestEncryptionLUKS1` triggers an assert
```
2024-07-19T12:48:08.399 INFO:tasks.ceph.osd.3.smithi110.stderr:/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos9/DIST/centos9/MACHINE_SIZE/gigantic/release/19.0.0-5126-g01e1aa05/rpm/el9/BUILD/ceph-19.0.0-5126-g01e1aa05/src/osd/ECCommon.cc: In function 'static void ECCommon::ReadPipeline::get_min_want_to_read_shards(uint64_t, uint64_t, const ECUtil::stripe_info_t&, const std::vector<int>&, std::set<int>*)' thread 7f9b60945640 time 2024-07-19T12:48:08.398035+0000
2024-07-19T12:48:08.399 INFO:tasks.ceph.osd.3.smithi110.stderr:/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos9/DIST/centos9/MACHINE_SIZE/gigantic/release/19.0.0-5126-g01e1aa05/rpm/el9/BUILD/ceph-19.0.0-5126-g01e1aa05/src/osd/ECCommon.cc: 336: FAILED ceph_assert(want_to_read->size() == sinfo.get_data_chunk_count())
2024-07-19T12:48:08.401 INFO:tasks.ceph.osd.3.smithi110.stderr: ceph version 19.0.0-5126-g01e1aa05 (01e1aa052c15e2a077a20de402d3ae763de4a30d) squid (dev)
2024-07-19T12:48:08.401 INFO:tasks.ceph.osd.3.smithi110.stderr: 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x11e) [0x56123c9a347e]
2024-07-19T12:48:08.401 INFO:tasks.ceph.osd.3.smithi110.stderr: 2: ceph-osd(+0x3fd63a) [0x56123c9a363a]
2024-07-19T12:48:08.401 INFO:tasks.ceph.osd.3.smithi110.stderr: 3: ceph-osd(+0x3a4a88) [0x56123c94aa88]
2024-07-19T12:48:08.401 INFO:tasks.ceph.osd.3.smithi110.stderr: 4: (ECCommon::ReadPipeline::get_min_want_to_read_shards(unsigned long, unsigned long, std::set<int, std::less<int>, std::allocator<int> >*)+0x56) [0x56123cc9b886]
2024-07-19T12:48:08.401 INFO:tasks.ceph.osd.3.smithi110.stderr: 5: (ECCommon::ReadPipeline::objects_read_and_reconstruct(std::map<hobject_t, std::__cxx11::list<ECCommon::ec_align_t, std::allocator<ECCommon::ec_align_t> >, std::less<hobject_t>, std::allocator<std::pair<hobject_t const, std::__cxx11::list<ECCommon::ec_align_t, std::allocator<ECCommon::ec_align_t> > > > > const&, bool, std::unique_ptr<GenContext<std::map<hobject_t, ECCommon::ec_extent_t, std::less<hobject_t>, std::allocator<std::pair<hobject_t const, ECCommon::ec_extent_t> > >&&>, std::default_delete<GenContext<std::map<hobject_t, ECCommon::ec_extent_t, std::less<hobject_t>, std::allocator<std::pair<hobject_t const, ECCommon::ec_extent_t> > >&&> > >&&)+0x8f7) [0x56123cca48a7]
2024-07-19T12:48:08.401 INFO:tasks.ceph.osd.3.smithi110.stderr: 6: (ECBackend::objects_read_async(hobject_t const&, std::__cxx11::list<std::pair<ECCommon::ec_align_t, std::pair<ceph::buffer::v15_2_0::list*, Context*> >, std::allocator<std::pair<ECCommon::ec_align_t, std::pair<ceph::buffer::v15_2_0::list*, Context*> > > > const&, Context*, bool)+0x5a0) [0x56123ceb8240]
2024-07-19T12:48:08.401 INFO:tasks.ceph.osd.3.smithi110.stderr: 7: (PrimaryLogPG::OpContext::start_async_reads(PrimaryLogPG*)+0x179) [0x56123cbda069]
2024-07-19T12:48:08.401 INFO:tasks.ceph.osd.3.smithi110.stderr: 8: (PrimaryLogPG::execute_ctx(PrimaryLogPG::OpContext*)+0x474) [0x56123cc0adc4]
2024-07-19T12:48:08.402 INFO:tasks.ceph.osd.3.smithi110.stderr: 9: (PrimaryLogPG::do_op(boost::intrusive_ptr<OpRequest>&)+0x342b) [0x56123cc036ab]
2024-07-19T12:48:08.402 INFO:tasks.ceph.osd.3.smithi110.stderr: 10: (OSD::dequeue_op(boost::intrusive_ptr<PG>, boost::intrusive_ptr<OpRequest>, ThreadPool::TPHandle&)+0x197) [0x56123cb41697]
2024-07-19T12:48:08.402 INFO:tasks.ceph.osd.3.smithi110.stderr: 11: (ceph::osd::scheduler::PGOpItem::run(OSD*, OSDShard*, boost::intrusive_ptr<PG>&, ThreadPool::TPHandle&)+0x69) [0x56123cd864d9]
2024-07-19T12:48:08.402 INFO:tasks.ceph.osd.3.smithi110.stderr: 12: (OSD::ShardedOpWQ::_process(unsigned int, ceph::heartbeat_handle_d*)+0xd07) [0x56123cb5ade7]
2024-07-19T12:48:08.402 INFO:tasks.ceph.osd.3.smithi110.stderr: 13: (ShardedThreadPool::shardedthreadpool_worker(unsigned int)+0x2aa) [0x56123d05ef6a]
2024-07-19T12:48:08.402 INFO:tasks.ceph.osd.3.smithi110.stderr: 14: ceph-osd(+0xab9514) [0x56123d05f514]
2024-07-19T12:48:08.402 INFO:tasks.ceph.osd.3.smithi110.stderr: 15: /lib64/libc.so.6(+0x89c02) [0x7f9b83a89c02]
2024-07-19T12:48:08.402 INFO:tasks.ceph.osd.3.smithi110.stderr: 16: /lib64/libc.so.6(+0x10ec40) [0x7f9b83b0ec40]
2024-07-19T12:48:08.402 INFO:tasks.ceph.osd.3.smithi110.stderr:*** Caught signal (Aborted) **
```
in the OSD were excluded from last the rerun.  This issue is being looked at by the RADOS team.)

idryomov commented:
jenkins test api

idryomov commented:
> @idryomov as you mentioned earlier, previous implementation of `group snap ls` command that displayed just the snap name and state would work when the pool of a group's image is removed from underneath. The new implementation of `group snap ls` introduced in this PR will fail in such a scenario (See https://tracker.ceph.com/issues/67021).

I don't think the fact that this PR makes the situation worse for `rbd group snap ls` command is a blocker.  As noted in the tracker, many other commands break when a pool is removed underneath a group and cleanup using just rbd CLI is impossible today.  To "recover" the group, one would need to reach into and manipulate the group metadata using rados CLI and they would be able to see the snapshots there.

> I guess we'll fix this properly when working on https://tracker.ceph.com/issues/67021 ?

Yes, when someone gets to it.

idryomov commented:
> Hey, friendly request here to disable new tests on crimson or pinging once `test_rbd.py` changes if possible. 

@Matan-B I'm not following how something related to this PR could be not supported by Crimson.  We simply exposed a few additional fields through the API here.

> It breaks Crimson suite once an unsupported test is added.

I think the issue with https://pulpito.ceph.com/matan-2024-07-24_07:30:51-crimson-rados-wip-sjust-crimson-testing-2024-07-22-distro-crimson-smithi/ is the mismatch between the branch that was built and the branch that was used to get the test definitions.

Matan-B commented:
> @Matan-B I'm not following how something related to this PR could be not supported by Crimson. We simply exposed a few additional fields through the API here.

> I think the issue with https://pulpito.ceph.com/matan-2024-07-24_07:30:51-crimson-rados-wip-sjust-crimson-testing-2024-07-22-distro-crimson-smithi/ is the mismatch between the branch that was built and the branch that was used to get the test definitions.

Ah, this is possible since test_rbd.py was pulled from main and the build's sha1 is older than that.
Thanks.
